File: Data Structures & Algorithms/trapping-rain-water/submission-2.js
class Solution {
    /**
     * @param {number[]} height
     * @return {number}
     */
    trap(height) {
        let res = 0;
        let start = 0, end = height.length - 1;
        let leftMax = height[0], rightMax = height[height.length - 1];

        while (start <= end) {
            if (leftMax < rightMax) {
                leftMax = Math.max(leftMax, height[start]);
                res += leftMax - height[start];
                start += 1;
            } else {
                rightMax = Math.max(rightMax, height[end]);
                res += rightMax - height[end];
                end -= 1;
            }
        }

        return res;
    }
}

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-2.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        # Find the indexes of the nearest shorter bars at each index
        # Use a strictly increasing stack to track the INDEX
        n = len(heights)
        leftStack, rightStack = [], []
        leftMax = [-1] * n
        rightMax = [n] * n

        for i in range(n):
            while leftStack and heights[leftStack[-1]] >= heights[i]:
                leftStack.pop()
            
            if leftStack:
                leftMax[i] = leftStack[-1]
            
            leftStack.append(i)

            while rightStack and heights[rightStack[-1]] >= heights[n - 1 - i]:
                rightStack.pop()
            
            if rightStack:
                rightMax[n - 1 - i] = rightStack[-1]
            
            rightStack.append(n - 1 - i)
        
        maxArea = 0
        for i in range(n):
            # Use the current bar as the height
            leftMax[i] += 1
            rightMax[i] -= 1
            currentArea = heights[i] * (rightMax[i] - leftMax[i] + 1)
            maxArea = max(maxArea, currentArea)
        
        return maxArea
